### ISSUE TITLE
FOSDEM talk next week

### ISSUE THREAD

**Issue by bartoldeman** [stsp/dosemu2#555]
Hi,

as the mailing list is not so active I am abusing an issue for this request. Next week I will be at FOSDEM (also for other work-related reasons) and will talk about DOSEMU and FreeDOS:
https://fosdem.org/2018/schedule/event/dosemu_and_freedos_past_present_future/
I can talk a little bit about DOSEMU2. Apart from looking for highlights in the NEWS file, is there anything important you (Stas) would like me to mention?

Also, to others who are subscribed to the issues, any cool applications you know of DOSEMU?

By the way I haven't tried it yet with a Meltdown/KPTI enabled kernel but can't imagine those are very friendly to DOSEMU...

**Comment by stsp** [comment]
Why is this all happens while I am
working on a freedos fork? :) I am sure
there would be much more interesting
things to talk about, but its not there yet.
Just to give you some numbers, I have
currently:
```
94 files changed, 3651 insertions(+), 1487 deletions(-)
```
And this is of course not counting the
white-space noise of this PR:
https://github.com/PerditionC/fdkernel/pull/14
which would otherwise give a diffstat of
```
147 files changed, 5199 insertions(+), 3078 deletions(-)
```
But don't worry, I will not call it "freedos2",
the name which is mentioned as your fosdem topic.

As for dosemu2 itself, the highlights can be
extracted from here:
https://github.com/stsp/dosemu2/releases
or from NEWS.
The things that are not mentioned in NEWS/releases,
are: the fully rewritten video stack, much improved
sound stack, mostly rewritten sighandling and DPMI
context switching, better EMS support (with paging of
lowmem).
One thing that definitely deserves a talk, is a way
I now support win32 under HX extender. There is
a famous problem with mapping win32 binaries at
a fixed address that may be unavailable because
of the memory layout randomization. Wine worked
around this by AFAIK writing the custom ld.so -
wineloader. dosemu2 solves this much better: I
just reserve some mem after lowmem itself, and
this reserve is always available for DPMI linear
mappings. Thanks to your non-zero mem_base
trick, this costed nearly nothing. So we have a
fully functional relocation machinery that is based
only on LDT segment base! No custom preloaders!
I even enabled pie, and it still works.

There are a few commercial deployments of dosemu2
that I am aware of. Namely, it is used in a POS
terminals with some commercial MUMPS interpreter:
https://en.wikipedia.org/wiki/MUMPS
For that I had to add an advanced EMM support,
as this interpreter has the built-in multitasker and
memory pager. dosbox doesn't support that, but
it probably can be easily added there.
Another application was some saw controller on
a furniture factory. For that I wrote a support for
custom serial port boards. Also I think @bolle732 
uses it commercially for running GEOS:
https://en.wikipedia.org/wiki/GEOS_(16-bit_operating_system)
for which we had many VESA modes emulation fixes
as well as the EMM fixes.

Essentially I think dosbox _could_ adopt these
things too, but it refused to do so, as it is only for
games. One thing it can't adopt though, is speed.
dosemu2 is much faster, except for 2 cases:
PIT emulation is bad, so progs that calculate
delays based on PIT counters sampling (rather
than based on PIT interrupts) are horribly slow.
Plus I've got a very slow emulation of unchained
VGA mode (compared to both dosemu1 and dosbox).

> By the way I haven't tried it yet with a Meltdown/KPTI
> enabled kernel but can't imagine those are very friendly to DOSEMU...

Forget it. :) They are not enabled on AMD processors
by default any longer, as AMD processors are immune to this.
And future Intel CPUs will likely be too. Plus, the meltdown
threat is exaggerated by miles. It is not even nearly
as big as the media claims it is. In a nutshell, it allows
you to steal only the memory that is already in an L1 cache -
not more, not less. The reliably ways of prefetching an
arbitrary kernel memory into L1 cache are yet to be seen.

**Comment by bartoldeman** [comment]
Thanks for the comments. I've got quite a bit to work with now combined with the ML reactions.

As for FreeDOS 2.0, it's just a proposal/discussion for a FreeDOS distribution that succeeds FreeDOS 1.2, not the kernel, see https://sourceforge.net/p/freedos/mailman/message/36080115/

Looking at your github are you talking about your gcc branch that probably works with plain gcc  (using protected mode?) instead of needing tkchia's ia16-elf-gcc branch right?

I've so far managed to compile the kernel (changed merged into github) and freecom (still on svn) with ia16-elf-gcc. Generated code not as small as Open Watcom yet, but passable (slightly better than Turbo/Borland).

**Comment by stsp** [comment]
Yes, that's the basic idea.
I want a fully protected mode DOS.
Of course this may not happen as
its widely accepted as impossible
wrt the freedos codebase. But was
this ever the reason not to try. :)

**Comment by stsp** [comment]
Hmm, oh, a Retrocumputing devroom on fosdem...
Time flies too quickly.
Maybe if we code up a prot mode DOS, then next
time you can qualify for an Embedded devroom
or Virtualization devroom instead? :)

**Comment by stsp** [comment]
> Regarding PC/GEOS:
> Yes, it's currently still a commercial product. The new
> owner plans to make as much as possible open source.

You mean, open-source geos itself, or something else?

> Like others that run Windows 3.11 under DOSEmu(2).

Correction: Windows 3.1 under dosemu2.
Other combinations (3.11 under dosemu1 etc) are not viable.

Thanks for the good feed-back!

**Comment by stsp** [comment]
> And in this regards, KVM is really the way into the future.

I don't think this is technically true.
KVM offers a lot of machinery to accelerate
the ring-0 system code, but dosemu doesn't
use those. For example, we don't need
nested page tables, which is the biggest
thing of KVM. For our needs the good jit
engine could do similar or even better better job.
Unfortunately, writing the good jit engines
is not something we have the resources for.
So KVM is the way into the future, but its
not the best possible one, as being not portable.
Of course its great that we have it right now.
Maybe one day there will be a good jit around
for us to steal. :)

**Comment by bolle732** [comment]
> You mean, open-source geos itself, or something else?

Yes, almost any parts of kernel, drivers and application should be open sourced. They are still reviewing the source code regarding third party licenses like the font system or thesaurus which can't be open sourced. But as we all know, open source is not equal to GPLx or any other free license. So that would be the interesting statement I'm still waiting for...

> Correction: Windows 3.1 under dosemu2.
> Other combinations (3.11 under dosemu1 etc) are not viable.

Thank you for the correction. I haven't done a lot with Windows 3.x.

**Comment by jschwartzenberg** [comment]
Damn, next week already I should've paid more attention to FOSDEM. If there'll be a dosemu talk again next year I'll definitely come to Brussels.

**Comment by andrewbird** [comment]
Perhaps it will be videoed?

**Comment by stsp** [comment]
https://video.fosdem.org/2018/

**Comment by stsp** [comment]
Another interesting thing for fosdem may
be to run win95 under dosemu2, like this:

![2018-01-31 03-07-32](https://user-images.githubusercontent.com/3295627/35598373-1662ce6c-0634-11e8-89a7-6e0c5e2a7c80.png)

**Comment by stuaxo** [comment]
Wow, when did this start working?

**Comment by bartoldeman** [comment]
I can show the screenshot, but what's Calimera exactly? Google did not find much.

**Comment by stsp** [comment]
Hi Bart!

Hope I am not too late.
I wanted to demonstrate the boot-up
screen of my freedos fork till fosdem,
so I worked this week for the whole nights. :)
And it finally appeared!
For the moment this can only be compiled
with clang as gcc seems to produce bad
code. But its here, see below:
![2018-02-03 20-44-18](https://user-images.githubusercontent.com/3295627/35769893-415cf0d8-0923-11e8-9722-9357b1f6f7de.png)

So what we see, besides the "hello world"
msg here? Well, we see that I can use both
gdb and dosdebug for development. Namely,
dosdebug here shows the trace through the
`INTR` func of freedos (I changed it to be FAR
as I can't jump to NEAR funcs from prot mode).
At gdb window we see the call stack of `init_call_intr()`.
Hopefully with such instrumentation the development
will go fluently. Unfortunately gdb is quite buggy,
it has problems with debugging fdpp, but basic
debugging functionality works.

Well, so while this is all I have for now, I hope
this proves that the entire thing is not a joke.
While you are talking about "past, present and future
of FreeDOS", please remember that I am here
doing everything possible to hijack this future
into the completely unpredictable and unexpected
routes. :)

**Comment by stsp** [comment]
Filled in a gcc bug:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=84194
Once it is fixed, fdpp can be compiled with g++
(it compiles even now, but the resulting code
doesn't work).
Its actually amazing that clang does the
better job than gcc these days.
The only problem with clang I've had during
fdpp development, was this one:
https://bugs.llvm.org/show_bug.cgi?id=36131
and even that appears not a clang's bug but
a C++ standard defect, which clang guys
already reported to the committee.
So it really seems clang is getting more friendly
than gcc in many aspects: better support from
community, fewer bugs, better diagnostic.
One of the clang contributors, @AntonBikineev,
was actually helping me to write this crazy
stuff, patiently explaining the advanced
meta-programming tricks with C++ templates.
Definitely not something to expect from gcc. :)

**Comment by bartoldeman** [comment]
Thanks. I added it to the talk at the end:
https://docs.google.com/presentation/d/1bC60lAA-0AhUOqbpe9-uFxazGBOwnFGpqs-bCmU0uDM/edit?usp=sharing
these slides were based on Jim Hall's for the FreeDOS part.
Note that I only have 30 minutes which typically flies, with an audience who may not necessarily all know what DPMI means... I really can't go into much detail.

**Comment by stsp** [comment]
Nice, thanks!
Let me also point to dosbox-x:
http://dosbox-x.com/
I haven't tried it myself, but it seems
to be having some really huge developments.

Also I find this statement a bit upsetting:
```
For example, a few folks wanted to create a “32-bit” FreeDOS that might break compatibility, but add a bunch of neat features. I’ll talk about a thought experiment for what a “modern DOS” might look like - it’s Linux!
```
Firstly, this "few folks" now also include myself,
and secondly, freedos-32 (http://freedos-32.sourceforge.net/)
is not a someone's wishful thoughts. It is a HUGE
source of code and info. The amount and the quality
of their code is really impressive: they have almost
every line commented, every sub-project carefully
documented, etc. I'll be stealing the code from them
as time permits, starting from their command.com:
https://github.com/stsp/comcom32
Its quite sad that this project have ended after
so many work was done (I think it was actually
commercial, and probably ran out of business).
Now I am trying to achieve in a few month what
they tried to achieve in 7 years (project was active
from 2000 to 2007). So this is an interesting
challenge. :)

**Comment by stsp** [comment]
> with an audience who may not necessarily all know what DPMI means...
> I really can't go into much detail.

Please note that fdpp project is not about DPMI,
its all about meta-programming. It consists of
bison-written syntax analysers, sed/awk transformers,
C++ templates that analyse the farpointer-inferred data
flow inside freedos code, but has almost no code that
contributes to the resulting binary. This is the only
hope why this project may succeed within the limited
amount of time I can spend on it: I really hope that a
few month of meta-programming can give more than
years of C/asm development. :)
But this remains to be seen.

**Comment by bartoldeman** [comment]
sure, those comments are Jim Hall's old comments from his presentation here: http://www.freedos.org/jhall/kieler/, I won't be presenting it that way myself (I probably won't have the time in any case!). They're also not part of the final presentation (which is a pdf file).

**Comment by stsp** [comment]
http://mirror.onet.pl/pub/mirrors/video.fosdem.org/2018/AW1.121/dosemu_and_freedos_past_present_future.mp4

Didn't know Hans Lermen passed away. :(
You seem to have a mistake there: `/proc/self/exe`
is supposed to be `/proc/self/mem`.
Another mistake is that HX actually allows you to
run opengl games, not only text-mode utilities.
I run quake2 with sound and very smooth rendering.
I guess some quake2 screenies could be good to
relief this very technical talk. :)
Oh, and there was almost no talk about the freedos
future, except for a couple of words about the upcoming
freedos2 distro release, which is mostly a present,
not really a future. So I am still in wonders of what
the future of freedos is. :)

**Comment by bartoldeman** [comment]
Thanks for the corrections. Guess I was too tired and forgetful when I wrote the exe instead of mem...
With HX I ran out of time to properly investigate it. My bad.

Actually there is very little known about the future of FreeDOS, just a little proposal and discussion on the mailing list.

Anyway was happy to see a full room, there were about 90 seats but another ~50 (I guess) people were standing and sitting on stairs. About 2/3 of those were dosemu users when I asked for the show of hands. 

Still nothing compared to the final plenary from the Red Hat's Jon Masters about Meltdown & Spectre... at least 1500 people.

**Comment by partoneoftwo** [comment]
Compare and contrast Dosbox with Dosemu2! At least mention how Dosemu2 is different! :) I have super fond memories of using this software, and it did work really well. I will take it for another spin soon.

**Comment by andrewbird** [comment]
@stsp Thanks for the video link.

**Comment by stsp** [comment]
> Actually there is very little known about the future of FreeDOS

I think you take this too literally, and the
talk was too technical as the result.
FreeDOS is all about the software it allows
to run. So showing the evolution of the
software that can be run on freedos may
be the interesting topic for next year.
And it will obviously make the speech much
less technical, allowing to talk more about future.
There are also other free DOSes around:
https://bitbucket.org/ecm/rxdos-7.2x/src
https://github.com/roelandjansen/pcmos386v501
EDR-DOS (seems dead).
All are supported by dosemu2, thanks to heroic
efforts of @andrewbird. :)
And there are notebooks still selling with freedos.
So I am sure we can gather enough info for any
talk about the future of FreeDOS(es) & its software,
but it will require much more time than 1 week.
So my proposal for next year is to come here,
say, 3 month before the actual presentation, and
set up something for collaborative work on the
presentation. People will likely help with bringing
their screenshots and stories, and you'll only edit
them into a final form. :)

**Comment by stsp** [comment]
By the way, would be cool if you can contribute
the linker changes you mentioned on fosdem, to
my project as well, so that I can get rid of those
jwlib/jwlink. This will allow me to concentrate
entirely on the compiler part of the problem.

**Comment by stsp** [comment]
There appears to be the C++ support library for djgpp:
https://github.com/jwt27/libjwdpmi
which looks quite active with commits.
Japhets is still there? :)

**Comment by stsp** [comment]
Some googling suggests that the present of
freedos is also quite interesting:
http://www.freedos.org/download/
Come here and the the full iso img worth of 400Mb!
What could be there that large? :)
Or come here:
https://www.gnu.org/software/gnudos/
and see something interesting too.
Next fosdem definitely have topics to speak. :)

**Comment by severach** [comment]
I'd be happy with an LFN MOVE.EXE. The one from DOS 7 works fine but I can't package that.

**Comment by jschwartzenberg** [comment]
The video was great to watch! @bartoldeman, thanks a lot for the talk! I think the audience was relatively young though. My question would've been "Would using KVM for protected mode/DPMI as well enable running Windows 95?" :)

**Comment by bolle732** [comment]
> The video was great to watch!

I only wished the camera was pointing a little bit more up. Unfortunately, the upper most part from the projection was cut off...

**Comment by bartoldeman** [comment]
Yes me too, many of the other FOSDEM videos have the slides big and the speaker in a corner but for some reason that did not happen with the setup in the retrocomputing devroom. However as a workaround you could follow with the slide pdf deck.

The slides are linked from this page:
https://fosdem.org/2018/schedule/event/dosemu_and_freedos_past_present_future/
https://fosdem.org/2018/schedule/event/dosemu_and_freedos_past_present_future/attachments/slides/2466/export/events/attachments/dosemu_and_freedos_past_present_future/slides/2466/dosemu_freedos.pdf

**Comment by bartoldeman** [comment]
@stsp I think most of my talk was not very technical but some parts were , and I think that's ok, some people in the audience are developers who know about V86 and LDTs and want to see a little detail.
In any case it's unlikely I'll be at FOSDEM again next year (it was just a lucky coincidence with work travel, and Brussels is 7 hours by plane from here), perhaps you should present then :) Although I realize that is easier said than done too.

What I referred to for linking the FD kernel is this linker script:
https://github.com/PerditionC/fdkernel/blob/master/kernel/kernel.ld
Mostly to get a bunch of ELF files linked together in a funny memory layout. OMF can do that from the object format but with ELF you cannot do segment relocations. The kernel.exe file has no relocations in its header but a "DGROUP" symbol (etc) is explicitly set by LD.

@jschwartzenberg Yes overall at FOSDEM you see a lot of 20-somethings but there were a few about 3x that age at the talk.

@jschwartzenberg in principle you could, though not in the patch I had. It's more work: either implement VCPI or run it in real mode as in the first KVM patches (but that breaks some other things in DOSEMU that depend on V86 trapping).

**Comment by stsp** [comment]
> @stsp I think most of my talk was not very technical but some parts were

What I mean by this, is that you mostly was
focusing on a technical challenges, and, at
large, on changelog. What I would like to hear
(I mean, I really don't know that material, so
I wanted to hear it from the talk) is what freedos
_actually_ is, what are those 400Mb of download,
what gnudos is, what are the benefits of selling
notebooks with freedos, what GUIs does freedos
provide, what dev tools, what internet tools
(arachne, not netscape! :), etc etc.
Changelog just can't be of any interest, and the
technical challenges definitely deserve some talk,
but not when as the result I still don't know what
freedos actually is. :) So its not any kind of bashing
or whatever, this is just what I expected to learn
from the video, but did not. And because of this,
I hope this chance will re-appear next year. So if
you happen to be there next year again and if you
agree that the topics I mentioned may be interesting,
then my proposal is to collaborate on them next time.

As for the @jschwartzenberg's question about
Windows95 - let me remind that we still don't know
(from the talk) if freedos even runs win95 and if so -
to what degree of compatibility. Since I suspect
it doesn't, I'd say neither VCPI nor realmode will
help on dosemu side.
Another point is that we have the reverse-engineered
JEMM, which was in the past one of the main points
for wanting VCPI, but now we can just support JEMM
directly. So as we can't run win95 with freedos anyway
and can (in principle) support JEMM, I don't see a
compelling case for VCPI right now, considering the
amount of efforts it will take even with the help of KVM.

**Comment by jschwartzenberg** [comment]
I would expect to need to use the DOS that ships with Windows 95 to run Windows 95 in dosemu. Even for the DR-DOS people it was very hard to write WinGlue/WinBolt.

**Comment by andrewbird** [comment]
I didn't know about those, so i looked and foumf this page. https://msfn.org/board/topic/109018-windows-98-in-dr-dos/

**Comment by stuaxo** [comment]
I don't think WinGlue / Winbolt are available anywhere on the internet though ?

I don't know how long the threat of MS lawyers last, but maybe the author could release it ?

**Comment by stsp** [comment]
While it might be nice to run win95
under dosemu, the reality is that today
under wine you can run much more
progs than under win95.
Also win4lin shows that it might be
simpler to re-implement the win95's
ring0 kernel stubs, rather than to create
the compatible DOS. In which case it
can be running under dosemu.

**Comment by haenschen** [comment]
Very interesting talk, thanks for sharing!

**Comment by stsp** [comment]
After taking a few week's break for recovering
my life from this freedos-fork coding marathon,
I've finally got back to it. Its new home is here:
https://github.com/stsp/fdpp

It now boots up to
`Bad or missing Command Interpreter: ed Enter the full shell command line`
which looks more than promising.
I think it have reached the state where
the hardcore coders like Bart and Andrew
can start looking into it.
Of course I am not hoping someone is going to
be interested until it actually boots command.com,
but I hope it is quite close to that point.
I think the porting framework is mostly completed.
Now its time to keep porting, fix bugs and all
that boring stuff.

**Comment by stsp** [comment]
> I think the porting framework is mostly completed.

Spoke too soon...
After fixing a few bugs, it no longer gets to
loading command.com. :)

**Comment by stsp** [comment]
Just wanted to share a bit of an fdpp progress.
After a half of year of various crashes and crappy
garbage on screen, I've finally (for the first time
since the banner screen I created for fosdem)
got a sensible messages from freedos on the
screen. So it really leaves somewhere under
this horrible porting framework, just wasn't showing
up till today. :)

![2018-05-09 03-18-10](https://user-images.githubusercontent.com/3295627/39789322-cf4ac1f0-5337-11e8-9c03-8cbfe3a423af.png)

**Comment by andrewbird** [comment]
That's looking very good!

BTW in the last sentence, did autocorrect change `lives` into `leaves` as I don't get the proper sense of what you are saying?

**Comment by stsp** [comment]
What kind of autocorrect?
Fixed the comment.

**Comment by andrewbird** [comment]
:+1:

**Comment by stsp** [comment]
As BlockIO is mostly ported, here's the first
booting screen. Of course it doesn't complete
the boot process yet. As can be seen from the
screenie, it have some problems parsing config.sys.

![2018-07-15 04-48-34](https://user-images.githubusercontent.com/3295627/42729743-b512ce4e-87ea-11e8-908a-6dd69f534a9b.png)

**Comment by stsp** [comment]
No, that was a bad screen-shot: the config.sys
parser in freedos was just plain broken. Don't
tell me it ever worked at all (on any compiler)
after looking into this commit:
https://github.com/stsp/fdpp/commit/3aede50ef9bbbc4e1461e10a9d2d27b3f0731a76

![2018-07-15 19-29-14](https://user-images.githubusercontent.com/3295627/42735969-3192acac-8866-11e8-9a4d-4858b35efa65.png)

So basically config.sys works, device drivers load.
Still can't load command.com, but it looks like the
last obstacle. Doesn't anyone want to be a hero and
fix the last command.com loading bug? :)

**Comment by jschwartzenberg** [comment]
So FreeDOS could never parse any config.sys? :)

**Comment by stsp** [comment]
See the commit above - if you think with
this bug it still could, then I won't agree. :)

**Comment by jschwartzenberg** [comment]
So how could people ever use FreeDOS? Will your fixes end up in regular FreeDOS as well?

**Comment by stsp** [comment]
> So how could people ever use FreeDOS?

The thing is, I don't even know.
Maybe their compilers were plain wrong
to compile that into something that worked.
Or maybe they had some strange "read()"
routine that uppercases the text beforehand,
and so, as the result, strcaseequal() could
"a kind of work" because the toupper() macro
was expanding the left and right side similar
number of times, therefore comparing, say,
every 3-rd char, but that could still match.
Well, I have hard times even thinking out how
could it theoretically pretend to work.

> Will your fixes end up in regular FreeDOS as well?

Very unlikely, but this particular one should, or
freedos will not work at all from now on.

**Comment by stsp** [comment]
Argh, no, this is wrong. :)
Looking into the original code, there is
no such macro in this file (it is there in
another files, but with the safer use).
Will amend the commit's msg, sorry for
false alarm. :)

**Comment by andrewbird** [comment]
Wow that looks cool :smile: Presumably that's with a FAT hdimage file?

Have you thought about the redirector yet? I wonder if it's possible to do the majority of the work without the traditional redirector and only maintain CDS for those progs that look for it. Although this does sound like the track that DR-DOS went down in the early versions.

**Comment by stsp** [comment]
No, this is a redirected drive under dosemu2!
The only change from the default setup I did,
was to comment 3 lines in fdconfig.sys:
```
rem config.sys for DOSEMU + FreeDOS
;SWITCHES=/F
;DOS=UMB,HIGH
dosdata=umb
lastdrive=Z
files=40
stacks=0
buffers=10
device=e:\dosemu\umb.sys
;devicehigh=e:\dosemu\ems.sys
shellhigh=c:\command.com /e:1024 /p
```

Andrew, please, why didn't you just try that
yourself? Building fdpp is way much easier
that building freedos with ia16-gcc, and the
latter you already do.
But I haven't yet touched the redirector part,
so on the redirected partition you can't do
much more than to type "ver" like on that
screen-shot. :) It barely works with redirector,
but it does! Even despite the fact that I didn't
do any porting in that area.

> I wonder if it's possible to do the majority of the
> work without the traditional redirector and only
> maintain CDS for those progs that look for it.

I don't understand what you mean.
Freedos maintains the CDS regardless of the
redirector, and this is unlikely to change.
But whatever you mean, fdpp surely provides
you with an unlimited possibilities to experiment.

**Comment by stsp** [comment]
I have added a toplevel "make install"
now so that you don't have to do "cd kernel"
before doing "sudo make install". :)
I don't know how to make the build/install
process even easier, everyone can just
take a look! :)

**Comment by andrewbird** [comment]
I suppose I don't fully get what you are doing here. I thought that you were running the freedos kernel in 32 bit land as a plugin, is that correct? If that were the case I don't think there's any reason that the freedos plugin has to implement the redirector interface in the way it does now for example:

An application does int21 disk read, kernel checks CDS flags, sees it's a network drive, then does int2f to a redirector which happens to be its own. The redirector interface has LFN and size limitations as is.

I thought maybe the kernel plugin could generate the int21 response itself using whatever methods it likes.

**Comment by stsp** [comment]
> I thought that you were running the freedos kernel in 32 bit land as a plugin, is that correct?

64bit, of course. :)

> An application does int21 disk read, kernel checks CDS flags,
> sees it's a network drive, then does int2f to a redirector which happens to be its own.

What kind of its own?
The redirector is always in dosemu, no
matter what DOS is used. The DOS does
not have a redirector (only have the hooks
to use one), so it can't have any of "its own".

> I thought maybe the kernel plugin could generate the int21
> response itself using whatever methods it likes.

But that's exactly what happens: DOS
(when running as a plugin or not) generates
the int21 response itself, calling out to the
redirector if he finds this appropriate.
If you mean to implement some other
redirector API that won't have the limitations,
then this is definitely possible with freedos,
no matter the plain freedos or fdpp.

**Comment by andrewbird** [comment]
> What kind of its own?

I was considering dosemu + fdpp as the same thing, so fdpp calls int2f which is implemented in this case by dosemu.

> If you mean to implement some other redirector API that won't have the limitations

Yep that was my first thought, but I later I presumed that fdpp could call directly into dosemu2 code. But reading between the lines, that appears not to be possible.

So just to be clear, I'm not trying to be rude, but what is the advantage to running freedos as a plugin?

**Comment by stsp** [comment]
> I was considering dosemu + fdpp as the same thing

Same thing? :)
They are a bit different.

> so fdpp calls int2f which is implemented in this case by dosemu.

Sure as hell.
How is that differs from plain freedos?
Same thing! Both call to dosemu's int2f
for the redirector API.

> Yep that was my first thought, but I later I presumed that fdpp could call directly into dosemu2 code.
> But reading between the lines, that appears not to be possible.

Why does this matter? Directly or indirectly -
these terms are too vague and unimportant
here. You may think that plain freedos calls
to dosemu's int2f indirectly because it involves
the complex transition to prot mode. Maybe
with fdpp its a bit more "direct", but why does
this ever matter?

> So just to be clear, I'm not trying to be rude,
> but what is the advantage to running freedos as a plugin?

What is the advantages of having any other
program in 64bit, rather than in real-mode?
Both work! But easier to build (with clang),
debug (gdb, valgrind) and maintain. It doesn't
look like any obvious advantage from an
outside POV, but being a dosemu developer,
and looking into a competitors like dosbox,
didn't you ever thought it might be good to
have freedos which can be built for a modern
prot-mode environment by modern tools?
I am sure dosemu2 deserves something
more than a plain old realmode DOS. :)
I am sure the advantages will become obvious
only if we start using it, or will not - if we abandon
that idea and stick to plain freedos.
For now its just a huge experiment, but I do
not doubt in its potential.

So what fdpp is _not_:
- A "freedos on steroids" or something like that.
No, it is a plain freedos + a compiler framework
to build it for 64bit.
- a dosemu2 plugin. No, it have no relationship
to dosemu2 - you can run it on freedos-32 kernel
http://freedos-32.sourceforge.net/
or any other sufficiently DOS-oriented prot-mode
kernel. Of course I haven't tried that myself, as
my primary target is dosemu2. But when fdpp is
mature enough, I hope freedos-32 people will
try it on their kernel. fdpp plugin in dosemu2 is
just a small "example" of how to use fdpp.

**Comment by stsp** [comment]
So let me put it that way: the 64bit port
of any real-mode program does not add
any immediate advantages by itself.
They are added later.
And yet people do the 64bit ports of
their software. Because the practice
shows it usually pays back. :)

**Comment by andrewbird** [comment]
Okay, I understand the purpose now.
Did you need to make any incompatible changes to FreeDOS that will prevent it building the usual 'kernel.sys' using the normal toolchain? I wonder if Dosemu2 + fdpp can become a pleasant development environment for freedos.
Will fdpp always remain standalone, or will it be added to dosemu2 repository?

**Comment by stsp** [comment]
> Did you need to make any incompatible changes to FreeDOS that will prevent it building the usual
> 'kernel.sys' using the normal toolchain?

The build system is completely botched up.
But this can be fixed: the fdpp files can just
be moved to the separate dir with their own
makefile, and the original build system can
be restored in "fdkernel" dir.
Also the includes are all botched up.
Originally there were separate includes for
INIT_TEXT modules and HMA_TEXT modules,
and I screwed that up. Should be easy to restore.
I am sure I did a lot of other incompatible
changes too, but they were unintentional
and/or are done via macros. The macros
that are now missing for other compilers,
will have to be added to `portab.h`.
So there was no intention to break the
original compilation, but, as you may guess,
I couldn't keep it alive myself, either.
Overall, restoring the original build will likely
be a few days of work (or hours, if you
are very fast:).

> I wonder if Dosemu2 + fdpp can become a pleasant development environment for freedos.

This was the intention.

> Will fdpp always remain standalone, or will it be added to dosemu2 repository?

Like systemd? :)
Well, maybe it can be added as a git submodule.
Its too difficult to predict now. If people will start
getting problems with it being standalone, then
we can add. But for now I don't see such a need.
I have no problems compiling it in a different dir.

**Comment by andrewbird** [comment]
> Like systemd? :)

Hopefully not. It was more a case of if it became dosemu2's own special DOS, like Dosbox has.

**Comment by stsp** [comment]
> Hopefully not. It was more a case of if it became dosemu2's own special DOS, like Dosbox has.

So it is possible for the sake of the
build simplicity (as a git submodule),
but you have to make the following
choice. Do you want a:
- "pleasant development environment for freedos"
- "dosemu2's own special DOS, like Dosbox has"

Both quotes are yours, please choose whatever
you like more. :) I don't see them being compatible
to each other.

**Comment by andrewbird** [comment]
I suppose it depends which feature is more successful.

**Comment by stsp** [comment]
I tagged the third alpha release of fdpp:
https://github.com/stsp/fdpp/releases/tag/alpha-3
This release is very stable and can boot
default config/autoexec of dosemu2, without
any mods to them. It can use HMA, loadhigh,
shellhigh and all that. Quite impressive.
As soon as the filesystem problems are solved,
we go beta.
I added an option to dosemu.conf that allows
to disable the redirector. This is for @andrewbird 
to play with: it appears fdpp has some problems
with FS even on non-redirected drives.
Note: w/o redirector, `dir` in fdpp works ~15 seconds
if built with `-O0`, and ~3 seconds for `-O2` build.
Which means something is very sub-optimal there,
but it doesn't hang, just some patience is needed.

**Comment by stsp** [comment]
![2018-07-22 03-31-20](https://user-images.githubusercontent.com/3295627/43041179-12d45a04-8d61-11e8-88d1-df36a64378aa.png)

**Comment by stsp** [comment]
https://github.com/stsp/fdpp/releases/tag/alpha-4
Alpha-4 of fdpp is released, which got a huge amount
of maintenance work on all fronts. Now fdpp is
available in dosemu2 PPA on launchpad:
https://launchpad.net/~dosemu2/+archive/ubuntu/ppa
And comcom32 now finally boots under it:
https://github.com/stsp/comcom32
And dosemu2 now loads fdpp by default, if available
(otherwise it loads freedos, as before).
Hopefully the maintenance work is all done, so I
am planning to fix the remaining problems and go
to Beta stage. Any help with that is much appreciated.
There is still the problem that comcom32 requires
djgpp, which is not packaged for ubuntu (but is
packaged for fedora). I wonder if porting comcom32
to clang (which will likely mean yet another meta-compiler
framework) worth the troubles. But really, fdpp doesn't
look complete without its own command.com in 64bit
space, so maybe I should port also that...
